### ISSUE TITLE
Q: Idiomatic way to do weight tying/sharing?

### ISSUE THREAD

**Issue by colehaus** [patrick-kidger/equinox#469]
A question rather than an issue.

It seems relatively straightforward to mimic the behavior of weight sharing by: `tree_at`ing or `tree_map`ing after initialization to copy initial weights over and then combining gradients to ensure they stay synchronized. But this, AIUI, would be mostly silly because you wouldn't reduce memory usage at all.

It in fact seems there is no realistic post-hoc `tree_map`, etc you can do to turn a non-shared architecture into one with shared weights (in a performance-relevant way)? Because a shared-weight model corresponds to a fundamentally different Pytree and there is no notion of shared reference/alias in the JAX idiom?

The best or only approach is to manually pass shared weights down to the pieces that need them at `__call__` time? e.g. if `Parent` has two children that are supposed to share weights, `Parent` could keep `weights` as a class attribute and pass it to each child when `__call__`ing?

I see some discussion of [how do this in Flax](https://github.com/google/flax/discussions/1264), but it looks like Flax maybe already implements things by registering and threading parameters around so it's more straightforward to do weight sharing in that approach?

(For a little more context, I'm implementing a version of the T5 paper and thinking about sharing weights across the encoder and decoder halves. But it looks like that might entail pretty substantive changes to the model architecture.)

**Comment by patrick-kidger** [comment]
So your approach of storing them on the parent,and passing them at `__call__` time, is the one I use and recommend.

> Because a shared-weight model corresponds to a fundamentally different Pytree and there is no notion of shared reference/alias in the JAX idiom?

Yup, exactly! At that point it's not a tree any more, but a DAG.

That said, at some point I'd like to see if I can find a better way of handling this. (Although I don't have any elegant solutions at the moment; I'd welcome any thoughts.)

You may also like `eqx.tree_check`, which can be used to check that you haven't accidentally got repeated nodes in the tree.

**Comment by colehaus** [comment]
Okay, that makes sense. Thanks!

**Comment by colehaus** [comment]
It's a little gross but preliminary tests suggest that something like this works as a way to avoid duplicating the core `__call__` logic for each `Module` (once in the non-shared version and once in the shared version):

```py
A = TypeVar("A", bound=Callable[..., Any])


class FakeCallable(Generic[A]):
    def __new__(cls, *args: Any, **kwargs: Any) -> A:
        return super().__new__(cls)  # pyright: ignore

    def __init__(self, entries: Mapping[str, Any], call: type[A]):
        self.__dict__.update(entries)
        self._call = call.__call__  # pyright: ignore

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        return self._call(self, *args, **kwargs)


class LinearCall(TypedDict, Generic[InDim, OutDim, Float]):
    """The attributes of `eqx.nn.Linear` used during `__call__`"""
    in_features: InDim
    out_features: OutDim
    weight: ndarray[OutDim, InDim, Float]
    bias: ndarray[OutDim, Float]


def fake_linear(x: LinearCall[InDim, OutDim, Float]) -> eqx.nn.Linear[InDim, OutDim, Float]:
    return FakeCallable(x, eqx.nn.Linear)
```

Then a call site is just:

```py
        fake_layer = fake_linear(
            {
                "in_features": self.embed_size,
                "out_features": self.intermediate_size,
                "weight": layer_params.weight,
                "bias": layer_params.bias,
            }
        )
        fake_layer(inputs) # Instead of self.layer(inputs)
```

**Comment by patrick-kidger** [comment]
I don't understand? It looks like your code resolves to just `eqx.nn.Linear.__call__(LinearCall(...), inputs)`. Moreover this all appears to occur at `__call__` time, but I think the problem we're trying to tackle is how to represent a DAG at `__init__` time.

**Comment by patrick-kidger** [comment]
Ah, I see. So the details vary depending on what you're doing, but the "pass to call" approach is usually done by just instantiating the sublayers inside `__call__`:
```python
class LinearAddOne(eqx.Module):
    linear: eqx.nn.Linear

    def __call__(self, x):
        return self.linear(x) + 1

class MyModule(eqx.Module):
    shared_layer: eqx.nn.Linear

    def __init__(self, key):
        self.shared_layer = eqx.nn.Linear(2, 3, key=key)

    def __call__(self, x):
        sublayer1 = SubLayer(self.shared_layer)
        sublayer2 = SubLayer(self.shared_layer)
        return sublayer1(x) + sublayer2(x)
```
This does require that `SubLayer` have no other parameters, though. (Else they're not part of the pytree!) In those cases then some chicanery with `eqx.tree_at` can be useful. Extract the shared layer at `__init__` time and then re-insert it at `__call__` time.

I appreciate that none of this is ideal. I've got some ideas noodling around for how this might be improved; let me have a think.

**Comment by colehaus** [comment]
Ah, yeah, that's certainly more straightforward for the common case where the layer type is also the same at each site. I was thinking of cases where you have different behavior associated with the parameters like sharing weights across the embedder and final softmax layer.

No worries at all! I'm definitely a fan of the library!

**Comment by colehaus** [comment]
Okay, I managed to look at this a bit.

- Unless I'm much mistaken, the typical way weights are shared between the embedding layer and the final pre-softmax layer is that the weight matrix is transposed. Is this new functionality intended to support that use case? If doesn't seem like it does currently due to the `tree_equal` check on line 65.
- When I tried this, with an `eqx.filter_jit` function that calls an `eqx.filter_value_and_grad(has_aux=True)` function, I get this error from JAX: 

  > TypeError: Cannot interpret value of type <class 'equinox.nn._shared._ShareIndex'> as an abstract array; it does not have a dtype attribute

  which is thrown in [this line](https://github.com/google/jax/blob/292deef6fda9f639fcecd9883a1112825a1eb54f/jax/_src/api_util.py#L574).

  I haven't really dug in to figure out whether this is user error, but I can give more info and reproducing code if that would be useful.

(Also, when trying to setup weight sharing to test this, I had code like this (following the [compilation speed trick](https://docs.kidger.site/equinox/tricks/#improve-compilation-speed-with-scan-over-layers)):

```py
class LM(eqx.Module, Generic[VocabSize, Float]):
    …

    def __init__(…):
        …
        def mk_self_attention(key: KeyArray) -> SelfAttention[LM.EmbedSize, Float]:
            …

        self.self_attention = eqx.filter_vmap(mk_self_attention)(
            to_aux_dim(jax.random.split(self_attn_key, num=num_layers))
        )

    def __call__(…) -> ndarray[OutSeqLen, VocabSize, Float]:
        …
        encoder_layers: jax.AuxDim[LM.NumLayers, EncoderLayerPartialApply[LM.EmbedSize, Float]] = eqx.filter_vmap(
            EncoderLayerPartialApply
        )(self.self_attention, self.feed_forward)
```
But the call to `filter_vmap` produces:

> ValueError: vmap was requested to map its argument along axis 0, which implies that its rank should be at least 1, but is only 0 (its shape is ())

I think because `self.self_attention` is a `Tracer` within `__call__`? Is that behavior expected? Is there a good way to do the `vmap` fast compilation trick in this setting? I ended up having to resort to lists and for loops.)

**Comment by patrick-kidger** [comment]
Thanks for giving this a go!

- Ah, good point about transposition. Hmm, in general it seems like we'd need a function to transform the weight in some arbitrary way before putting it back in the pytree. In the transposition case it'd be `jnp.transpose`.

    I'll need to think about what the cleanest API is for this -- we somehow need to express (a) the paths to multiple nodes via lambdas, (b) which node in a group is kept, and (c) an optional transformation, that defaults to the identity function. (Happy to take suggestions.)
    
- `filter_jit`: can you give a MWE?

- `filter_vmap`: it should definitely be possible to use a vmap here. `self.self_attention` should be a pytree, which includes tracers. Again, do you have a MWE? (Also, are you intending to use vmap or scan?)

**Comment by colehaus** [comment]
Ahh, I figured out that `jax.checkpoint` is part of the problem in both cases.

Here's the MWE for the new `Shared` functionality:

```py
import equinox as eqx
import jax
import jax.numpy as jnp
from jax.numpy import ndarray


@jax.checkpoint
def loss(model) -> ndarray[float]:
    return jnp.array(0.0)


class MyModule(eqx.Module):
    shared: eqx.nn.Shared

    def __init__(self, key):
        key1, key2 = jax.random.split(key)
        embedding = eqx.nn.Embedding(num_embeddings=3, embedding_size=4, key=key1)
        head = eqx.nn.Linear(4, 3, key=key2)

        sharing = lambda m: (m[0].weight, m[1].weight)
        self.shared = eqx.nn.Shared((embedding, head), sharing)

if __name__ == "__main__":
    loss(MyModule(jax.random.PRNGKey(0)))
```

and here's for `filter_vmap`:

```py
import equinox as eqx
import jax
import jax.numpy as jnp

class Dropout(eqx.Module):
    p: float

    def __init__(self, p: float = 0.5):
        self.p = p

class DropoutStatic(eqx.Module):
    p: float = eqx.field(static=True)

    def __init__(self, p: float = 0.5):
        self.p = p

class MyModel(eqx.Module):
    layers: eqx.Module

    def __init__(self):
        self.layers = eqx.filter_vmap(lambda _: Dropout())(jnp.arange(4)) # Error
        # self.layers = eqx.filter_vmap(lambda _: DropoutStatic())(jnp.arange(4)) # No error

    def __call__(self):
        encoder_layers = eqx.filter_vmap(lambda x: None)(self.layers)


@jax.checkpoint
def fun1(m):
    m()


if __name__ == "__main__":
    fun1(MyModel())
```

But I guess it makes sense that these don't work out of the box? There would need to be something like `eqx.filter_checkpoint`?

**Comment by patrick-kidger** [comment]
Ah, yes. This is indeed an issue with `jax.checkpoint`.

To explain what's going on. The first MWE fails fairly straightforwardly: because `MyModule(jax.random.PRNGKey(0))` is a PyTree of both JAX arrays and non-JAX-arrays. And `jax.checkpoint` doesn't know how to handle non-JAX-arrays. For example, you can get the same error just by doing
```python
mlp = eqx.nn.MLP(2, 2, 2, 2, key=key)
jax.checkpoint(lambda f: None)(mlp)
```

The second is a little more subtle: in this case, the `filter_vmap` in `__init__` doesn't actually do anything: there are no JAX arrays in the output, just a `Dropout(p=0.5)`. Then when `MyModel()` is passed across the `jax.checkpoint`, JAX upcasts the `0.5` into an `Array(0.5)`. Then the `filter_vmap` in `__call__` detects the (scalar) array, and tries to vmap across it.

---

Basically, it's unfortunate that JAX tries to cast arraylikes into arrays -- this introduces quite a few subtle issues like the one above. (Conversely, Equinox *has* to filter out arraylikes, because otherwise statements like `if foo:` will fail with a concretisation error.)

It sounds like it might be worth adding an `eqx.filter_checkpoint`. I'd be happy to take a PR on this.

**Comment by dlwh** [comment]
> It sounds like it might be worth adding an eqx.filter_checkpoint. I'd be happy to take a PR on this.

I actually have an implementation of this. I will try to PR tonight or tomorrow.

I'll try to make it a bit better, but here it is in the meantime:

```
def filter_checkpoint(fun: Callable, *, prevent_cse: bool = True, policy: Optional[Callable[..., bool]] = None):
    """As `jax.checkpoint`, but allows any Python object as inputs and outputs"""

    @ft.wraps(fun)
    def _fn(_static, _dynamic):
        _args, _kwargs = eqx.combine(_static, _dynamic)
        _out = fun(*_args, **_kwargs)
        _dynamic_out, _static_out = eqx.partition(_out, is_jax_array_like)
        return _dynamic_out, Static(_static_out)

    checkpointed_fun = jax.checkpoint(_fn, prevent_cse=prevent_cse, policy=policy, static_argnums=(0,))

    @ft.wraps(fun)
    def wrapper(*args, **kwargs):
        dynamic, static = eqx.partition((args, kwargs), is_jax_array_like)
        dynamic_out, static_out = checkpointed_fun(static, dynamic)

        return eqx.combine(dynamic_out, static_out.value)

    return wrapper
```

**Comment by dlwh** [comment]
(i'd probably replace the hard-coded filter predicate with something configurable, as one does)

**Comment by patrick-kidger** [comment]
Thank you, that'd be great! I think the only detail I'd suggest adjusting is to make the return value be a PyTree -- typically an `eqx.Module` subclass. C.f. something like `equinox.filter_value_and_grad` for an example.

**Comment by colehaus** [comment]
Ah, that makes sense. Thanks all! I looked and the reason I didn't initially identify `jax.checkpoint` as the culprit is that I had been using it fine, but it turns out every part of my `Module` just so happened be array-like!

**Comment by patrick-kidger** [comment]
Regarding transposition: I'm considering an API of the form
```python
attn = eqx.nn.MultiheadAttention(...)
lin = eqx.nn.Linear(...)
source = lambda attn, lin: attn.key_proj.weight
dest = lambda attn, lin: lin.weight
self.shared = eqx.nn.Shared(attn, lin).share(source, dest, transform=jnp.transpose)
# weird sharing between the key projection weight and a linear weight, just to demo
```
where `transform=lambda x: x` by default, and multiple shares can be handled by doing `Shared(...).share(...).share(...)` arbitrarily many times.

It's still a little bit of an odd API, but it seems to meet the basic requirements of (a) offering an optional transform, (b) offering the ability to share multiple things at a time, (c) make it clear which direction the transform is being applied in, from `source` to `dest`, (d) avoiding the weird `pair[0]` and `pair[1]` business that the previous implementation had.

Definitely open to alternative ideas on what would feel like a natural API.

**Comment by colehaus** [comment]
This API looks pretty reasonable overall to me. Some thoughts:

1. Should `Shared` explicitly take two arguments like this? Or just a single pytree? If I had a pre-existing architecture/nested `Module`, it seems like it would be nice to be able to just pass that directly to `Shared`.

```py
module = SomeComplicatedModule(…)
self.shared = eqx.nn.Shared(module).share(
  lambda m: m.encoder.layers[0].feed_forward.weight, 
  lambda m: m.decoder.layers[0].feed_forward.weight
).share(
  lambda m: m.encoder.layers[1].feed_forward.weight, 
  lambda m: m.decoder.layers[1].feed_forward.weight
)
```

2. As the preceding example suggests, I can imagine it being nice for `share` to take a sequence of (named?) tuples instead of expecting repeated `share` calls. It seems like the sharing structure may often be data-like. With the current approach, I think it might be common for people to have to write a helper like:

```py
def share_many(shared, shares):
    for s in shares:
        shared = shared.share(*s)
    return shared
```

Obviously very possible and not too onerous. But which API is more fluent probably depends on predicted use patterns (more often one-off sharings or more often multiple, structured shares) which are maybe hard to know in advance. 

3. If it's going to take two explicit arguments like this, do each of `source` and `dest` need to be given two arguments? Or can it be simplified to:
```py
source = lambda attn: attn.key_proj_weight
dest = lambda lin: lin.weight
 ```

**Comment by patrick-kidger** [comment]
(1)/(3) Ah, so what I'm suggesting is for `Shared(*args)` to then call `source(*args)` and `dest(*args)`. Thus `Shared(pytree).share(lambda p: ...)` is valid, as is `Shared(foo, bar).share(lambda f, b: ...)`.

(2) I considered something like this, but I don't see how to make this happen without complicating the API in the simple case. We need to be able to optionally pass a transform for each destination, after all.

I won't deny this approach does still annoy me, just because I'd prefer something where we just initialise a single `Shared` object -- which is morally speaking all we want -- rather than needing to chain a bunch of `.share` calls off the end.

**Comment by colehaus** [comment]
1/3. Okay, thought that might be possible. Just wanted to clarify.
2. You could theoretically use a `NamedTuple` to do the defaulting on the transform, but that doesn't seems like a very Pythonic API. Something like:

```py
class ShareArgs(NamedTuple, Generic[*T, A, B]):
    source: Callable[[*T], A]
    target: Callable[[*T], B]
    transform: Optional[Callable[[A], B]] = None


class Shared(Generic[*T]):
    def __init__(self, *args: *T) -> None:
        ...

    # What we really want here is an existential type rather than `Any` but alas
    def share(self, args: list[ShareArgs[*T, Any, Any]] | ShareArgs[*T, Any, Any]):
        ...

x = eqx.nn.Linear(in_features=2, out_features=3, key=jax.random.PRNGKey(0))
y = eqx.nn.Linear(in_features=3, out_features=2, key=jax.random.PRNGKey(0))

Shared(x, x).share(ShareArgs(lambda x, _: x.weight, lambda _, y: y.weight))
Shared((x, y)).share(
    [
        ShareArgs(lambda x: x[0].weight, lambda x: x[1].weight, transform=jnp.transpose),
        ShareArgs(lambda x: x[0].bias, lambda x: x[1].bias),
    ]
)
```

**Comment by patrick-kidger** [comment]
Alright, thanks for the conversation! I think I've figured this one out. The API is `eqx.nn.Shared(pytree, where, get)`. See that PR for details -- I think the result is both readable and general.

WDYT?

**Comment by colehaus** [comment]
I think the motive for weight sharing is often reducing memory usage in memory-constrained environments. I may be misunderstanding something about `tree_at` or the cleverness of the compilation pass, but it looks to me like there's no real memory savings at `__call__` time?

**Comment by patrick-kidger** [comment]
Here's a quick numpy analogy:
```python
import equinox as eqx
import numpy as np

### Init time

def init_linear():
    weight = np.random.random(size=(1000, 1000))
    bias = np.random.random(size=(1000,))
    return (weight, bias)  # pytree with weight and bias

def init_embedding():
    weight = np.random.random(size=(1000, 1000))
    return (weight,)  # pytree with weight

def init_model():
    linear = init_linear()
    embedding = init_embedding()
    embedding = eqx.tree_at(lambda e: e[0], embedding, None)  # deduplicate: remove extra weight
    return (linear, embedding)  # pytree with linear and embedding

model = init_model()
# At this point we only have two arrays in memory: the linear weight and bias. The
# embedding's weight has been deleted.

### Call time

def call(model, ...):
    linear, embedding = model
    (weight, _) = linear
    embedding = eqx.tree_at(lambda e: e[0], weight, weight)
    # At this point, we happen to be able to access the weight in two different ways:
    # both linear[0] and embedding[0]. But it's the same Python object, and we only have
    # a single array in memory.
    ...

call(model, ...)
```
Basically, `tree_at` just rearranges the Python-level pytrees, but it doesn't make any array copies. Here, it allows us to find the underlying JAX array in two different ways, either through the linear layer or through the embedding layer.

FWIW, it seems like we could maybe do with better docs if this is a point of confusion. (Either for `tree_at` or for `Shared`, or both.) Do you know what it would take to clarify this for you?

**Comment by colehaus** [comment]
Ah, okay. I was definitely misunderstanding `tree_at`. I don't see anything in the documentation for `tree_at` or `Shared` that strongly suggests my erroneous interpretation so it may just be an idiosyncratic misunderstanding. Possible tweaks:

"At  `__call__` time, the nodes at `where(pytree)` will be filled with the value of `get(pytree)`." → "At `__call__` time, the nodes at `where(pytree)` will be filled with references to the value of `get(pytree)`." (for `Shared`)

And:

"The return value from `replace_fn` will be used in its place." → "The nodes specified by `where` will be updated with references to the return value from `replace_fn`." (for `tree_at`)

(Note that it looks like this part of the `Shared` documentation is currently missing a space after `__call__`.)

**Comment by colehaus** [comment]
I'm using this now and it seems to be working well. Thanks for all the help!
